File: Episode024/WebStencilsDemos/WebBrokerProject/C++/UtilsSearch.cpp
//---------------------------------------------------------------------------
#include <vcl.h>
#pragma hdrstop

#include "UtilsSearch.h"
#include "UtilsLogger.h"

//---------------------------------------------------------------------------
#pragma package(smart_init)

//---------------------------------------------------------------------------
// TSearchParams Implementation
//---------------------------------------------------------------------------
__fastcall TSearchParams::TSearchParams()
    : System::Classes::TPersistent()
{
    FSearchTerm = "";
    FHasSearch = false;
}

//---------------------------------------------------------------------------
__fastcall TSearchParams::TSearchParams(const String& ASearchTerm, const std::vector<String>& ASearchFields)
    : System::Classes::TPersistent()
{
    FSearchTerm = ASearchTerm.Trim();
    FSearchFields = ASearchFields;
    UpdateHasSearch();
}

//---------------------------------------------------------------------------
__fastcall TSearchParams::~TSearchParams()
{
    FSearchFields.clear();
}

//---------------------------------------------------------------------------
void TSearchParams::UpdateHasSearch()
{
    FHasSearch = !FSearchTerm.IsEmpty() && !FSearchFields.empty();
}

//---------------------------------------------------------------------------
String TSearchParams::GetSearchSQL()
{
    if (!FHasSearch)
        return "";
        
    try
    {
        String SearchSQL = " WHERE (";
        
        for (size_t I = 0; I < FSearchFields.size(); I++)
        {
            if (I > 0)
                SearchSQL += " OR ";
                
            SearchSQL += "UPPER(" + FSearchFields[I] + ") LIKE UPPER(:search)";
        }
        
        SearchSQL += ")";
        
        return SearchSQL;
    }
    catch (Exception &E)
    {
        Logger->Error(String("Error generating search SQL: ") + E.Message);
        return "";
    }
}

//---------------------------------------------------------------------------
void TSearchParams::SetSearchParameters(TFDQuery* AQuery)
{
    if (!AQuery || !FHasSearch)
        return;
        
    try
    {
        // Set the search parameter with wildcards
        String SearchValue = "%" + FSearchTerm + "%";
        AQuery->ParamByName("search")->AsString = SearchValue;
        
    }
    catch (Exception &E)
    {
        Logger->Error(String("Error setting search parameters: ") + E.Message);
    }
}

//---------------------------------------------------------------------------
// TBaseSearch Implementation
//---------------------------------------------------------------------------
__fastcall TBaseSearch::TBaseSearch(const std::vector<String>& ASearchableFields)
    : TObject()
{
    FSearchableFields = ASearchableFields;
    
}

//---------------------------------------------------------------------------
__fastcall TBaseSearch::~TBaseSearch()
{
    FSearchableFields.clear();
}

//---------------------------------------------------------------------------
std::unique_ptr<TSearchParams> TBaseSearch::CreateSearchParams(TWebRequest* ARequest)
{
    if (!ARequest)
        return std::make_unique<TSearchParams>();
        
    try
    {
        // Get search term from query parameters
        String SearchTerm = "";
        
        // Try different common parameter names
        if (ARequest->QueryFields->IndexOfName("search") >= 0)
            SearchTerm = ARequest->QueryFields->Values["search"];
        else if (ARequest->QueryFields->IndexOfName("q") >= 0)
            SearchTerm = ARequest->QueryFields->Values["q"];
        else if (ARequest->QueryFields->IndexOfName("term") >= 0)
            SearchTerm = ARequest->QueryFields->Values["term"];
            
        SearchTerm = SearchTerm.Trim();
        
        auto SearchParams = std::make_unique<TSearchParams>(SearchTerm, FSearchableFields);
        return SearchParams;
    }
    catch (Exception &E)
    {
        Logger->Error(String("Error creating search params from request: ") + E.Message);
        return std::make_unique<TSearchParams>();
    }
}

//---------------------------------------------------------------------------
std::unique_ptr<TSearchParams> TBaseSearch::CreateSearchParams(const String& ASearchTerm)
{
    try
    {
        auto SearchParams = std::make_unique<TSearchParams>(ASearchTerm, FSearchableFields);
        return SearchParams;
    }
    catch (Exception &E)
    {
        Logger->Error(String("Error creating search params: ") + E.Message);
        return std::make_unique<TSearchParams>();
    }
}

File: Episode024/WebStencilsDemos/WebBrokerProject/Delphi/Utils.Search.pas
unit Utils.Search;

interface

uses
  Web.HTTPApp,
  FireDAC.Comp.Client,
  Utils.Logger;

type
  // Base search class that can be used for any entity
  TBaseSearch = class
  private
    FSearchFields: TArray<string>;
  public
    constructor Create(const ASearchFields: TArray<string>);
    function GetSearchFields: TArray<string>;
    function GetSearchSQL(const ASearchTerm: string): string;
    function GetDefaultSearchFields: string;
  end;

  // Search parameters class
  TSearchParams = class
  private
    FSearchTerm: string;
    FBaseSearch: TBaseSearch;
  public
    constructor Create(ARequest: TWebRequest; ABaseSearch: TBaseSearch);
    function GetSearchSQL: string;
    function GetSearchTermForUrl: string;
    function GetHasSearch: Boolean;
  published
    property HasSearch: Boolean read GetHasSearch;
    property SearchTerm: string read FSearchTerm;
    property BaseSearch: TBaseSearch read FBaseSearch;
  end;

implementation

uses
  System.SysUtils,
  System.NetEncoding;

{ TBaseSearch }

constructor TBaseSearch.Create(const ASearchFields: TArray<string>);
begin
  FSearchFields := ASearchFields;
end;

function TBaseSearch.GetSearchFields: TArray<string>;
begin
  Result := FSearchFields;
end;

function TBaseSearch.GetDefaultSearchFields: string;
var
  Field: string;
begin
  Result := '';
  for Field in FSearchFields do
  begin
    if Result <> '' then
      Result := Result + ',';
    Result := Result + Field;
  end;
end;

function TBaseSearch.GetSearchSQL(const ASearchTerm: string): string;
var
  Field: string;
  WhereClause: string;
begin
  if ASearchTerm = '' then
  begin
    Result := '';
    Exit;
  end;

  try
    WhereClause := '';
    
    for Field in FSearchFields do
    begin
      if WhereClause <> '' then
        WhereClause := WhereClause + ' OR ';
      WhereClause := WhereClause + Format('UPPER(%s) LIKE UPPER(:search)', [Field]);
    end;
    
    Result := 'WHERE (' + WhereClause + ')';
  except
    on E: Exception do
    begin
      Logger.Error(Format('Error generating search SQL: %s', [E.Message]));
      Result := '';
    end;
  end;
end;

{ TSearchParams }

constructor TSearchParams.Create(ARequest: TWebRequest; ABaseSearch: TBaseSearch);
begin
  try
    FBaseSearch := ABaseSearch;
    FSearchTerm := Trim(ARequest.QueryFields.Values['search']);
  except
    on E: Exception do
      Logger.Error(Format('Error creating search params from request: %s', [E.Message]));
  end;
end;

function TSearchParams.GetHasSearch: Boolean;
begin
  Result := FSearchTerm <> '';
end;

function TSearchParams.GetSearchSQL: string;
begin
  if not HasSearch or not Assigned(FBaseSearch) then
  begin
    Result := '';
    Exit;
  end;
  
  Result := FBaseSearch.GetSearchSQL(FSearchTerm);
end;

function TSearchParams.GetSearchTermForUrl: string;
begin
  if HasSearch then
    Result := TNetEncoding.URL.Encode(FSearchTerm)
  else
    Result := '';
end;

end. 